src/main/kotlin/org/ageseries/libage/sim/thermal/Thermal.kt

```diff
@@ -0,0 +1,250 @@
+) {
+    val contactPoint = params.contactPoint
+    val distance = params.distance
+    val conductance = params.conductance
+    /**
+     * Returns the change in energy, added to [a] and subtracted from [b], which would equilibriate the connected thermal masses over the given period of time [dt] (in s).
+     *
+     * For this to be stable, [dt] must be held relatively small, since this is a linear approximation to an exponential curve--otherwise, overshoot may be observed.
+     */
+    fun transfer(dt: Double): Double {
+        // Kelvin
+        val delta_T = b.temperature.kelvin - a.temperature.kelvin
+        // W/K
+        val dist_cond_a = a.material.conductance * contactPoint * distance
+        val dist_cond_b = b.material.conductance * (1.0 - contactPoint) * distance
+        val overall_cond = (dist_cond_a * dist_cond_b * conductance).pow(1.0/3.0)
+        // W
+        val power = delta_T * overall_cond
+        // J
+        return power * dt
+    }
+}
+
+interface ThermalBody<Locator> {
+    val mass: ThermalMass
+    val locator: Locator
+    val surfaceArea: Double
+}
+
+interface Environment<Locator> {
+    fun temperature(locator: Locator): Temperature
+    fun conductance(locator: Locator): Double
+}
+
+class Simulator<Locator>(val environment: Environment<Locator>) {
+    class Connection<Locator>(
+        val a: ThermalBody<Locator>,
+        val b: ThermalBody<Locator>,
+        params: ThermalConnectionParameters,
+    ) {
+        val connection = ThermalConnection(a.mass, b.mass, params)
+    }
+
+    val connections = mutableSetOf<Connection<Locator>>()
+    val bodies = mutableSetOf<ThermalBody<Locator>>()
+    val connectionMap = mutableMultiMapOf<ThermalBody<Locator>, Connection<Locator>>()
+
+    fun add(vararg bodies_added: ThermalBody<Locator>) {
+        bodies.addAll(bodies_added)
+    }
+
+    fun add(vararg connections_added: Connection<Locator>) {
+        connections.addAll(connections_added)
+        connections_added.forEach {
+            add(it.a, it.b)
+            connectionMap[it.a] = it
+            connectionMap[it.b] = it
+        }
+    }
+
+    fun connect(a: ThermalBody<Locator>, b: ThermalBody<Locator>, params: ThermalConnectionParameters = ThermalConnectionParameters.DEFAULT) {
+        add(Connection(a, b, params))
+    }
+
+    fun remove(vararg bodies_removed: ThermalBody<Locator>) {
+        bodies.removeAll(bodies_removed)
+        bodies_removed.forEach { body ->
+            // toList() here is just used to make a cheap copy
+            connectionMap[body].toList().forEach {
+                remove(it)
+            }
+        }
+    }
+
+    fun remove(vararg connections_removed: Connection<Locator>) {
+        connections_removed.forEach {
+            connectionMap[it.a].remove(it)
+            connectionMap[it.b].remove(it)
+        }
+    }
+
+    fun step(dt: Double) {
+        val delta_E = mutableMapOf<ThermalBody<Locator>, Double>()
```

Comment:
Can this be cached? We'll be doing tens of thousands of steps per second, so caching it seems sensible.

Reply:
`deltaE`

Reply:
Are you proposing caching the mutable map lookup, or the entire step function? Caching the map is probably not worth it, since the search time of a map is as fast as searching a cache would be, unless the indexing variable is excessively processing heavy compared to a cache alias.

Reply:
The mutable map lookup. Tens of thousands of moderately sized instances are created per second. These are allocations that can easily be avoided.